\begin{abstract}
DNA aligning is a compute-intensive and time-consuming task required for all further DNA processing. It consists in finding for each DNA string from a sample its location in a reference genome. Usual techniques for short reads (hundreds of bases) involves seed-extension, where a small matching seed is found with quick search through FM-index, and then extended on both ends with a custom Smith-Waterman algorithm, giving optimal solution. However, this seed-extension takes a tremendous amount of time; which is why we present in this thesis a solution to offload extension on a GPU to be done in a parallel fashion. This is possible thanks to the fact that the DNA reads do not present any kind of relation between each other. We used the Burrows-Wheeler Aligner (BWA), a reference program in the field, to which we integrated a GPU-accelerated library for extension, GASAL2. However, BWA has a left-right dependency on extension starting scores, with the left alignment starting with the seed score, then the right part starting with the previously calculated left score. We designed a decent solution by starting both extensions with the seed score, something we called "seed-only" paradigm. On our test machine featuring 12 hyperthreaded cores and an NVIDIA Tesla K40c, when running with 12 threads, we could observe a raw kernel speed-up of 4.8$\times$ ; but if we allow non-blocking calls to let the CPU run the seeding while the GPU does the extension, we can reach a 16$\times$ effective speed-up. This extension part taking 27\% of the total times but it introduces a small overhead due to memory preparations and copying, which makes the whole application 1.28$\times$ faster, getting close to the theoretical maximum of 1.37$\times$. Additionally, the paradigm shift we operated creates a negligible difference in the final overall scores, with a 0.23\% difference for our 5.2 million sequences data set. This makes our acceleration with GASAL2 an efficient solution for a single machine.
\end{abstract}
